solidity/contracts/bridge/library/FraudVerification.sol

```diff
@@ -0,0 +1,285 @@
+
+        uint256 challengeKey =
+            uint256(
+                keccak256(abi.encodePacked(walletPublicKey, sighash, v, r, s))
+            );
+
+        require(
+            data.fraudChallenges[challengeKey].reportedAt > 0,
+            "Fraud not challenged"
+        );
+
+        require(
+            !data.fraudChallenges[challengeKey].defended,
+            "Fraud challenge already defended"
+        );
+
+        uint32 sighashType = extractSighashType(preimage);
+        require(sighashType == 1, "Wrong sighash type");
+
+        if (witness) {
+            verifyWitnessPreimage(preimage, deposits, spentMainUTXOs);
+        } else {
+            verifyNonWitnessPreimage(preimage, deposits, spentMainUTXOs);
+        }
+        // If we passed the preimage verification, the wallet has successfully
+        // defended the fraud challenge.
+
+        // TODO: Reward the wallet for successful fraud challenge response.
+
+        data.fraudChallenges[challengeKey].defended = true;
+
+        // TODO: Consider emitting the event with walletPublicKey in the
+        //       compressed format as it's how we identify wallets in the Bridge.
+        emit FraudChallengeDefended(walletPublicKey, sighash, v, r, s);
+    }
+
+    // TODO: description
+    function verifyWitnessPreimage(
+        bytes memory preimage,
+        mapping(uint256 => Bridge.DepositRequest) storage deposits,
+        mapping(uint256 => bool) storage spentMainUTXOs
+    ) internal {
+        // A preimage created for a witness input contains the outpoint located
+        // at a constant offset of 68.
+        bytes32 previousOutpointTxIdLe = preimage.extractInputTxIdLeAt(68);
+        uint32 previousOutpointIndex =
+            BTCUtils.reverseUint32(uint32(preimage.extractTxIndexLeAt(68)));
+
+        uint256 utxoKey =
+            uint256(
+                keccak256(
+                    abi.encodePacked(
+                        previousOutpointTxIdLe,
+                        previousOutpointIndex
+                    )
+                )
+            );
+
+        // Check that the UTXO is among the correctly spent UTXOs.
+        require(
+            deposits[utxoKey].sweptAt > 0 || spentMainUTXOs[utxoKey],
+            "Spent UTXO not found among correctly spent UTXOs"
+        );
+    }
+
+    // TODO: description
+    function verifyNonWitnessPreimage(
+        bytes memory preimage,
+        mapping(uint256 => Bridge.DepositRequest) storage deposits,
+        mapping(uint256 => bool) storage spentMainUTXOs
+    ) internal {
+        // A preimage created during signing of a non-witness input contains
+        // all the inputs of the transaction. The input the preimage was
+        // generated for can be distinguished from other inputs by looking at
+        // unlocking scripts. The unlocking script of such input is non-null
+        // while unlocking scripts of other inputs are null.
```

Comment:
Yes, I confirmed that experimentally.
If we have a transaction with multiple inputs (some of them non-witness, some witness) and we are signing a non-witness input, the preimage generated for that input will contain all the inputs (including witness ones) and only one of them (the one being signed) will have a non-zero `scriptSig`.
The others will have `0x00` as their `scriptSig`.

But we don't have to relay on that.
We can simply iterate over all the inputs and check whether the outpoint (hash + index) for each input is marked in the Bridge as correctly spent UTXO (meaning it can be found in `deposits` or `spentMainUTXOs`).

There are three transactions a wallet can make:
- `sweep` - can have multiple inputs (all of them must be either in `deposits` or `spentMainUTXOs` to be considered correct)
- `redemption` - can only have one input (it must be in `spentMainUTXOs`)
- `transfer funds` - can only have one input (it must be in `spentMainUTXOs`).

So if we are verifying a preimage, we can check whether all the inputs found in that preimage are correctly spent in the Bridge.

Reply:
Shouldn't we just check the input that is under challenge?

Reply:
It would work both ways (we could check only the input that is challenged or we could check all the inputs from that transaction - they all should be proven in the Bridge).
But I change the logic so that only the input under challenge is checked in
85576ccb3f4704db8f82b117d6a6cf92de060a58.